File: rust/tools/test_framework/python_tests/stress_test.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from rclpy.qos import QoSProfile, ReliabilityPolicy, HistoryPolicy, DurabilityPolicy
import time, sys, argparse, json

try:
    from robonix_sdk.srv import SubmitTask
except ImportError:
    print("Error: robonix_sdk not found.")
    sys.exit(1)

def main():
    parser = argparse.ArgumentParser()
    parser.add_argument('--client-id', type=int, default=0)
    parser.add_argument('--requests', type=int, default=1000)
    parser.add_argument('--rate', type=int, default=100)
    parser.add_argument('--duration', type=int, default=0)
    args = parser.parse_args()
    
    rclpy.init()
    node = Node(f'stress_test_python_{args.client_id}')
    qos = QoSProfile(reliability=ReliabilityPolicy.RELIABLE, history=HistoryPolicy.KEEP_LAST, depth=10, durability=DurabilityPolicy.VOLATILE)
    client = node.create_client(SubmitTask, '/rbnx/task/submit', qos_profile=qos)
    
    if not client.wait_for_service(timeout_sec=10.0):
        print("Service not available")
        return

    stats = {'success': 0, 'failed': 0, 'latencies': []}
    interval = 1.0 / args.rate if args.rate > 0 else 0
    start_time = time.time()
    end_time = start_time + args.duration if args.duration > 0 else None
    last_req = start_time

    for i in range(args.requests):
        if (end_time and time.time() >= end_time): break
        elapsed = time.time() - last_req
        if elapsed < interval: time.sleep(interval - elapsed)
        last_req = time.time()
        
        req = SubmitTask.Request(description=f'task {i}', params='{}')
        call_start = time.time()
        
        future = client.call_async(req)
        while rclpy.ok() and not future.done() and (time.time() - call_start) < 5.0:
            rclpy.spin_once(node, timeout_sec=0.01)
            
        if future.done():
            try:
                future.result()
                stats['success'] += 1
                stats['latencies'].append((time.time() - call_start) * 1000.0)
            except:
                stats['failed'] += 1
        else:
            stats['failed'] += 1

    lats = sorted(stats['latencies'])
    total = stats['success'] + stats['failed']
    p = lambda pct: lats[int(len(lats) * pct / 100.0)] if lats else 0
    
    print(f"\nOverall Statistics:\nTotal: {total}, Success: {stats['success']}, Failed: {stats['failed']}")
    print(f"Latency (ms): Avg: {sum(lats)/len(lats) if lats else 0:.2f}, Min: {lats[0] if lats else 0:.2f}, Max: {lats[-1] if lats else 0:.2f}, P50: {p(50):.2f}, P95: {p(95):.2f}, P99: {p(99):.2f}, P999: {p(99.9):.2f}")
    
    node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__': main()

File: rust/tools/test_framework/cpp_tests/stress_test.cpp
#include <chrono>
#include <iostream>
#include <memory>
#include <string>
#include <vector>
#include <algorithm>
#include <numeric>

#include "rclcpp/rclcpp.hpp"
#include "robonix_sdk/srv/ping_pong.hpp"

using namespace std::chrono_literals;

class StressTestClient : public rclcpp::Node {
public:
  StressTestClient(int client_id)
      : Node("stress_test_cpp_" + std::to_string(client_id)) {
    client_ = this->create_client<robonix_sdk::srv::PingPong>("/rbnx/ping");
  }

  bool wait_for_service(std::chrono::seconds timeout) {
    return client_->wait_for_service(timeout);
  }

  std::shared_ptr<robonix_sdk::srv::PingPong::Response> call_service(int task_id, double& latency_us) {
    auto request = std::make_shared<robonix_sdk::srv::PingPong::Request>();
    request->message = "ping";
    request->sequence = task_id;

    auto start = std::chrono::high_resolution_clock::now();
    auto result_future = client_->async_send_request(request);

    // Wait for the result with a timeout
    if (rclcpp::spin_until_future_complete(this->get_node_base_interface(), result_future, 5s) ==
        rclcpp::FutureReturnCode::SUCCESS) {
      auto end = std::chrono::high_resolution_clock::now();
      latency_us = std::chrono::duration<double, std::micro>(end - start).count();
      return result_future.get();
    } else {
      return nullptr;
    }
  }

private:
  rclcpp::Client<robonix_sdk::srv::PingPong>::SharedPtr client_;
};

int main(int argc, char** argv) {
  int client_id = 0;
  int requests = 1000;
  int rate = 100;
  int duration = 0;

  for (int i = 1; i < argc; ++i) {
    std::string arg = argv[i];
    if (arg == "--client-id" && i + 1 < argc) client_id = std::stoi(argv[++i]);
    else if (arg == "--requests" && i + 1 < argc) requests = std::stoi(argv[++i]);
    else if (arg == "--rate" && i + 1 < argc) rate = std::stoi(argv[++i]);
    else if (arg == "--duration" && i + 1 < argc) duration = std::stoi(argv[++i]);
  }

  rclcpp::init(argc, argv);
  auto node = std::make_shared<StressTestClient>(client_id);

  if (!node->wait_for_service(10s)) {
    std::cerr << "Service not available" << std::endl;
    return 1;
  }

  std::vector<double> latencies;
  int success = 0;
  int failed = 0;
  
  auto start_time = std::chrono::high_resolution_clock::now();
  auto end_time = duration > 0 ? 
    start_time + std::chrono::seconds(duration) : 
    std::chrono::high_resolution_clock::time_point::max();
    
  double interval_s = rate > 0 ? 1.0 / rate : 0;
  auto last_req = std::chrono::high_resolution_clock::now();

  for (int i = 0; i < requests; ++i) {
    if (std::chrono::high_resolution_clock::now() >= end_time) break;
    
    auto now = std::chrono::high_resolution_clock::now();
    auto elapsed = std::chrono::duration<double>(now - last_req).count();
    if (elapsed < interval_s) {
      std::this_thread::sleep_for(std::chrono::duration<double>(interval_s - elapsed));
    }
    last_req = std::chrono::high_resolution_clock::now();

    double lat_us = 0;
    auto response = node->call_service(i, lat_us);
    if (response) {
      success++;
      latencies.push_back(lat_us);
    } else {
      failed++;
    }
  }

  std::sort(latencies.begin(), latencies.end());
  int total = success + failed;
  auto p = [&](double pct) {
    if (latencies.empty()) return 0.0;
    return latencies[static_cast<int>((latencies.size() - 1) * pct / 100.0)];
  };

  std::cout << "\nOverall Statistics:" << std::endl;
  std::cout << "Total: " << total << ", Success: " << success << ", Failed: " << failed << std::endl;
  
  double avg = latencies.empty() ? 0.0 : std::accumulate(latencies.begin(), latencies.end(), 0.0) / latencies.size();
  
  printf("Latency (us): Avg: %.2f, Min: %.2f, Max: %.2f, P50: %.2f, P95: %.2f, P99: %.2f, P999: %.2f\n",
         avg, p(0), p(100), p(50), p(95), p(99), p(99.9));

  rclcpp::shutdown();
  return 0;
}
